Design doc for thin logical volumes (#442)
topolvm/topolvm: docs/proposals/thin-volumes.md

## Proposal
- Topolvm is **not** responsible for creating the thin-pools. Storage admin/operator has to create the thin-pools in advance while creating the volume groups.
- Multiple device classes can be mapped to a single volume group having multiple thin pools. But each device class should map to a single thin pool within that volume group
- `lvmd.yaml` file should be updated to hold the mapping between device classes, volume group and thin pools
- For device classes mapped to thin pools, the node annotation`capacity.topolvm.cybozu.com/<device-class>` should be updated with respect to the thin pool size and overprovision ratio.
- Metrics related to thin volume `data` and `metadata` parameters should be exported.

## Design Details
- A storage admin has to create the volume groups and the thin-pools on this volume group.
- The storage admin should update the lvmd.yaml config file on the nodes to map device classes, volume groups and the corresponding thin-pools.
- Sample `lvmd.yaml` config file
``` yaml=
device-classes:
  - name: ssd-thin
    volume-group: myvg1
    default: true
    spare-gb: 10
    type: thin

>     thin-pool-config:

Author: Just in the lvmd.conf file?  Or do you want `DeviceClass.ThinPoolConfig` to be changed to `DeviceClass.ThinPool` in the API as well?

>     thin-pool-config:

Reviewer: - I feel it's only the json name, anyways let's wait for the response.

}
```
- Following new fields will be added to the API:
    - **Type**: Represents the type of the volume provisioning. Possible option right now is only `thin`. This field is optional. If no type is provided, then thick volumes should be provisioned.
    - **ThinPoolConfig**: Shows the configuration of thin pool used for creating thin volumes:
        - **Name**: Name of the thin pool
        - **OverProvisionRatio**: This represents the amount of overprovisioning allowed for the pool. For example: If the overprovisioningRatio is `5`, pool size is 100GiB, then this pool can be overprovisioned by only (5*100)500GiB.

#### Selecting Thin-pool
- The current behavior uses `topolvm.cybozu.com/device-class` parameter on the deviceClass to decide which volume group will be used.
- If the deviceClass has a thin-pool mapped to it, then thin volumes should be created using that thin-pool.
- If no thin-pool is mapped to the deviceClass, then thick volumes should be created.

#### LVMD service
- LVService reads the DeviceClass config in the lvmd.yaml.
- If the DeviceClass has thin-pool, `CreateLVRequest` should create thin volumes.

> - When creating thin logical volumes, the service should not compare the requested logical volume size with the available free size on the volume group. This is because thin pools allow overprovisioning.

Reviewer: - how about below as we'll not be comparing to the size of volume group when serving the request?

> [the maximum size is based on the size of the corresponding thin-pool] **than the available free size on the volume group.**
> [the maximum size is based on the size of the corresponding thin-pool] **with overprovision**

> - When creating thin logical volumes, the service should not compare the requested logical volume size with the available free size on the volume group. This is because thin pools allow overprovisioning.

Author: Update the description and used ideas provided by both. Let me know if that works.

- Command to create thin volumes:
``` go=
lvcreate -V 1G --thin -n thin_volume vg1/tpool
   WARNING: Sum of all thin volume sizes (1.00 GiB) exceeds the size of thin pool vg1/tpool and the size of whole volume group (992.00 MiB)!
   For thin pool auto extension activation/thin_pool_autoextend_threshold should be below 100.
   Logical volume “thin_volume” created.
```

#### LVMD Protocol changes

- For LVService
``` go=
message WatchItem {

>     uint64 free_bytes = 1;  // Free space of the volume group in bytes.

Reviewer: - how about setting free_bytes as overprovision_bytes if the device-class on the thin pool is set as default?

    string device_class = 2;
    uint64 size_bytes = 3;  // Size of the volume group in bytes.
    uint64 overprovision_bytes = 4; // Space available on thin pool with overprovision.
    float data_free = 5; // Data percent free in thin pool.

>     float metadata_free = 6; // Metadata percent free in thin pool.

Reviewer: - how about below, we'll take care of the corresponding changes to nodemetrics & monitormetrics.
``` proto
// ===== from ===== //
message WatchItem {
    uint64 free_bytes = 1;  // Free space of the volume group in bytes.
    string device_class = 2;
    uint64 size_bytes = 3;  // Size of the volume group in bytes.
}

// ===== to ===== //
// Represents the details of thinpool.
message ThinPoolItem {
  // Data and Metadata fields are used for monitoring thinpool
  float data_percent = 1; // Data percent occupied on the thinpool
  float metadata_percent = 2; // Metadata percent occupied on the thinpool

  // Used for scheduling decisions
  uint64 overprovision_bytes = 3; // Free space on the thinpool with overprovision.
}

// Represents the response corresponding to device class targets.
message WatchItem {
    uint64 free_bytes = 1; // Free space in the device class.
    string device_class = 2;
    uint64 size_bytes = 3; // Size of device class in bytes.
    ThinPoolItem thin_pool = 4;
}
```

>     float metadata_free = 6; // Metadata percent free in thin pool.

Reviewer: - general query, can I use data and metadata percent as `double` in `proto` so they'll be represented as `float64` in go and there'll be very few conversions?

>     float metadata_free = 6; // Metadata percent free in thin pool.

Reviewer 2: > general query, can I use data and metadata percent as double in proto so they'll be represented as float64 in go and there'll be very few conversions?

Is this question about gRPC than this proposal? If so, yes. `double` is mapped to `float64`.

https://developers.google.com/protocol-buffers/docs/proto3

In addition, I understood there is no ways to know the actual consumed space in bytes. Using percent unit is OK.

}
```
- Resize and remove of Logical Volumes will use the same API as the underlying calls on volume group or thinpool only needs Logical Volume ID.

#### Monitoring thin-pools
- We need to constantly monitor the thin-pools and ensure that it does not run out of space.
- `/etc/lvm/lvm.conf` settings can allow auto growth of the thin pool when required. By default, the threshold is 100% which means that the pool will not grow.
- If we set this to, 75%, the Thin Pool will autoextend when the pool is 75% full. It will increase by the default percentage of 20% if the value is not changed. We can see these settings using the command grep against the file.

```go=
# grep -E ‘^\s*thin_pool_auto’ /etc/lvm/lvm.conf
 thin_pool_autoextend_threshold = 100

>  thin_pool_autoextend_percent = 20

Reviewer: - We propose not to use dmeventd for monitoring thin pool and this mention is for the admin not to use these options (by default it's off)
- Instead we want the admins/users montoring stack generate alerts based on metrics that topolvm exports after this implementation

>  thin_pool_autoextend_percent = 20

Reviewer 2: Oops, I misunderstood. I read "/etc/lvm/lvm.conf" as "/etc/lvmd.conf". Sorry for noise.

- Note: Available thin pool on the volume groups will be obtained from lvmd.yaml config file.

## Open Questions
- Should topolvm create the thin-pools instead of the admin/operator?
    - Admin should create thin pools.

- If multiple thin-pools are supported for each volume group, then how to decide which thin-pool should be used to create the thin logical volumes.
    - We will have multiple thin pools per volume group. Multiple device classes can be mapped to a single volume group where each device class is mapped to a different thin pool within that volume group.

- How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.
    - Multiple device classes can be mapped to a single volume group where each device class is mapped to a different thin pool within that volume group. Capacity of the device class with thin pools can be calculated using the overprovision ratio.

- Should we expect auto-extend of thin-pool to be enabled by default? What happens when the user has already enabled the autoextend feature when doing the monitoring?

- Does topolvm constantly poll the available capacity of the thin pools?

> - How to reload lvmd to use updated lvmd.yaml?

Reviewer: - The operator is packaging lvmd with topolvm node plugin and we need to restart whole pod for lvmd to re-read the config
- Currently we are in search on how to do it intuitively

> - How to reload lvmd to use updated lvmd.yaml?

Reviewer 2: It's OK to find the way during implementation.

## Design Details
- A storage admin has to create the volume groups and the thin-pools on this volume group.
- The storage admin should update the lvmd.yaml config file on the nodes to map device classes, volume groups and the corresponding thin-pools.
- Sample `lvmd.yaml` config file:
``` yaml=
device-classes:
  - name: ssd-thin
    volume-group: myvg1
    spare-gb: 10
    type: thin
    thin-pool:
        name: pool0
        overprovision-ratio: 50.0
  - name: ssd
    volume-group: myvg1
    default: true
    spare-gb: 10
```
- Above example shows two device classes; `ssd-thin` and `ssd`. Both use `myvg1` volume group.
- `ssd-thin` has thin pool parameters.
- Only thin volumes will be created when `ssd-thin` device class is selected by the user in the storageClass parameter `capacity.topolvm.cybozu.com/device-class`
- Only thick volumes will be created when user has specified `ssd` in the storageClass.

#### API Changes

- The `DeviceClass` struct of the device-class-manager will hold the information about the thin-pool associated with the Volume Group.

``` go

+ type ThinPoolConfig struct {
+   // Name of thin pool
+   Name string`json:"name"`

> +   // OverProvisionRatio represents the ratio of overprovision that can be + allowed on thin pools

Reviewer 3: Unnecessary '+' is included.
```suggestion
+   // OverProvisionRatio represents the ratio of overprovision that can be allowed on thin pools
```

    - **ThinPoolConfig**: Contains the configuration of thin pool used for creating thin volumes. This field must exist if the `type` field is `thin` and ignored if the `type` field is not provided or the value is set to `thick`.
        - **Name**: Name of the thin pool
        - **OverProvisionRatio**: This represents the amount of overprovisioning allowed for the thin pool on the node. For example: If the overprovisioningRatio is `5`, pool size is 100GiB, then this pool can be overprovisioned by only (5*100)500GiB. This means that the sum of the virtual sizes of the thin volumes provisioned on the pool per node cannot cross 500GiB.

#### Selecting Thin-pool
- The current behavior uses `topolvm.cybozu.com/device-class` parameter on the StorageClass to decide which volume group will be used.
- If `DeviceClass.Type` is `thin`, then thin volumes should be created using `DeviceClass.ThinPoolConfig.Name` thin-pool.
- If `DeviceClass.Type` is `thick` or empty, then thick volumes should be created.

#### LVMD service
- LVService reads the DeviceClass config in the lvmd.yaml.

> - If`DeviceClass.Type` is `thin`, `CreateLVRequest` should create thin volumes using `DeviceClass.ThinPoolConfig.Name` thin-pool.

Reviewer 4: ```suggestion
- If `DeviceClass.Type` is `thin`, `CreateLVRequest` should create thin volumes using `DeviceClass.ThinPoolConfig.Name` thin-pool.
```

- When creating thin logical volumes, the service should not compare the requested logical volume size with the available free size on the volume group. This is because in case of thin volumes, the available free size is based on the size of the corresponding thin-pool (with overprovsioning) instead of the available free size on the volume group.
- Command to create thin volumes:
``` go=
lvcreate -V 1G --thin -n thin_volume vg1/tpool
   WARNING: Sum of all thin volume sizes (1.00 GiB) exceeds the size of thin pool vg1/tpool and the size of whole volume group (992.00 MiB)!
   For thin pool auto extension activation/thin_pool_autoextend_threshold should be below 100.
   Logical volume “thin_volume” created.
```

#### LVMD Protocol changes

- For LVService
``` go=
message WatchItem {
    uint64 free_bytes = 1;  // Free space of the volume group in bytes.
    string device_class = 2;
    uint64 size_bytes = 3;  // Size of the volume group in bytes.
    uint64 overprovision_bytes = 4; // Space available on thin pool with overprovision.
    float data_free = 5; // Data percent free in thin pool.
    float metadata_free = 6; // Metadata percent free in thin pool.

}
```

> - Resize and remove of Logical Volumes will use the same API as the underlying calls on volume group or thinpool only needs Logical Volume ID.

Reviewer 5: ```suggestion
- Resize and remove of Logical Volumes will use the existing API as the underlying calls do not require the thinpool information, only the Logical Volume ID.
```

- Multiple device classes can be mapped to a single volume group having multiple thin pools. But each device class should map to a single thin pool within that volume group.
- `lvmd.yaml` file should be updated to hold the mapping between device classes, volume group and thin pools.
- For device classes mapped to thin pools, the node annotation `capacity.topolvm.cybozu.com/<device-class>` should be updated with respect to the thin pool size and overprovision ratio.
- Metrics related to thin volume `data` and `metadata` parameters should be exported.

## Design Details
- A storage admin has to create the volume groups and the thin-pools on this volume group.
- The storage admin should update the lvmd.yaml config file on the nodes to map device classes, volume groups and the corresponding thin-pools.
- Sample `lvmd.yaml` config file:
``` yaml=
device-classes:
  - name: ssd-thin
    volume-group: myvg1
    default: true
    spare-gb: 10
    type: thin
    thin-pool:
        name: pool0
        overprovision-ratio: 50.0
  - name: ssd
    volume-group: myvg1

>     default: true

Reviewer 5: There are 2 entries mapping default to true in this example.

Author: fixed.

    spare-gb: 10
```
- Above example shows two device classes; `ssd-thin` and `ssd`. Both use `myvg1` volume group.
- `ssd-thin` has thin pool parameters.
- Only thin volumes will be created when `ssd-thin` device class is selected by the user in the storageClass parameter `capacity.topolvm.cybozu.com/device-class`
- Only thick volumes will be created when user has specified `ssd` in the storageClass.

#### API Changes

- The `DeviceClass` struct of the device-class-manager will hold the information about the thin-pool associated with the Volume Group.

``` go

+ type ThinPoolConfig struct {
+   // Name of thin pool
+   Name string`json:"name"`
+   // OverProvisionRatio represents the ratio of overprovision that can be + allowed on thin pools

> +   OverProvisionRatio float64 `json:"overprovision-ratio"`

Reviewer 5: What are the possible values for overProvisionRatio? Can it be set to 0 to disable it?

Author: I could think the possible values as `>0` and `<100`. 
By disabling overProvisionRatio, do you mean that no overprovision limit will be set and user can provision any number of thin logical volumes on a thin-pool?

I don't think we should give an option for the user to disable it for some thin-pools.
 - Can't think of a use case why user would want to disable it for some pools. 
 -  It will add more complexity.

Reviewer: possible values --> anything greater than 1.0
set to 0 to disable it --> what should be the annotation value if we set it to 0?

Reviewer 2: I forgot to answer this question. It's reasonable to set OverProvisionRatio to equal or greater than 1.0 and regards any other values as invalid because it's not "thin" provisioning.

+}

type DeviceClass struct {
    // Name for the device-class name
    Name string `json:"name"`
    // Volume group name for the device-class
    VolumeGroup string `json:"volume-group"`
    // Default is a flag to indicate whether the device-class is the default
    Default bool `json:"default"`
    // SpareGB is storage capacity in GiB to be spared
    SpareGB *uint64 `json:"spare-gb"`
    // Stripe is the number of stripes in the logical volume
    Stripe *uint `json:"stripe"`
    // StripeSize is the amount of data that is written to one device before moving to the next device
    StripeSize string `json:"stripe-size"`
    // LVCreateOptions are extra arguments to pass to lvcreate
    LVCreateOptions []string `json:"lvcreate-options"`
+   // Type of the volume to be created

> +   Type string `json:"type",omitempty`

Reviewer 5: What are the possible values of type ? Should this be an enum?

Reviewer: - new type LVType which coerces to string values `thick` or `thin`

Reviewer 4: Could you please remove `omitempty` option?
ref. https://github.com/topolvm/topolvm/pull/448#discussion_r834073781

Reviewer: sure, I'll get this design PR refreshed at the earliest.

+   // ThinPoolConfig represents the configuration options for creating thin volumes
+   ThinPoolConfig ThinPoolConfig `json:"thin-pool",omitempty`
}
```
- Following new fields will be added to the API:

>     - **Type**: Represents the type of the volume provisioning. Possible values are `thin` and `thick`. This field is optional. If no type is provided, then thick volumes should be provisioned.

Reviewer 5: ```suggestion
    - **Type**: Represents the type of the volume provisioning. Possible values are `thin` and `thick`. This field is optional. If no type is provided, it defaults to thick volumes.
```

>     - **ThinPoolConfig**: Shows the configuration of thin pool used for creating thin volumes. This field must exist if the `type` field is `thin` and ignored if the `type` field is `thick`.

Reviewer 5: ```suggestion
    - **ThinPoolConfig**: contains the configuration of thin pool used for creating thin volumes. This field must exist if the `type` field is `thin` and ignored if the `type` field is missing or the value is set to `thick` .
```

        - **Name**: Name of the thin pool

>         - **OverProvisionRatio**: This represents the amount of overprovisioning allowed for the pool. For example: If the overprovisioningRatio is `5`, pool size is 100GiB, then this pool can be overprovisioned by only (5*100)500GiB.

Reviewer 5: ```suggestion
        - **OverProvisionRatio**: This represents the amount of overprovisioning allowed for the thin pool on the node. For example: If the overprovisioningRatio is `5`, pool size is 100GiB, then this pool can be overprovisioned by only (5*100)500GiB. This means that the sum of the virtual sizes of the thin volumes provisioned on the pool per node cannot cross 500GiB.
```

#### Selecting Thin-pool

> - The current behavior uses `topolvm.cybozu.com/device-class` parameter on the deviceClass to decide which volume group will be used.

Reviewer 5: Please clarify this  - the annotation is not set on the deviceClass.

Author: Right.  Its on the StorageClass.  Updated.

- Command to be used for montoring:
``` go=
   lvs
   LV          VG   Attr       LSize   Pool  Origin Data%  Meta%  Move Log Cpy%Sync Convert
   lv1         vg1  -wi-ao---- 600.00m
   thin_volume vg1  Vwi-a-tz--   1.00g tpool        3.20
   tpool       vg1  twi-aotz-- 200.00m              16.38  1.37
```
- Topolvm will only monitor the thin-pools provided in the lvmd.yaml config file.

#### Capacity awareness
- Topolvm uses capacity awareness to ensure that new pod is created on the node where available capacity of the volume group is larger than the requested capacity. `capacity.topolvm.cybozu.com/<device-class>` annotation is added on each node which consists of the available capacity on the volume group.
- Creation of thin logical volumes doesn't update the available device class capacity until the user actually uses storage on the thin volumes.
- So we need to ensure that all the thin logical volumes don't get created on the same node.
- This can be done by using the `overprovisionRatio`
- For device classes having thin pools, the value of `capacity.topolvm.cybozu.com/<device-class>` annotation will be calculated as:

>   `(thin-pool-size * overprovisionedRatio)-(sum-of-all-thin-lv-sizes-in-thin-pool)`

Reviewer 5: ```suggestion
  `(thin-pool-size * overprovisionedRatio)-(sum-of-all-thin-lv-virtual-sizes-in-thin-pool)`
```

- Note: Available thin pool on the volume groups will be obtained from lvmd.yaml config file.

## Open Questions
- Should topolvm create the thin-pools instead of the admin/operator?
    - Admin should create thin pools.

- If multiple thin-pools are supported for each volume group, then how to decide which thin-pool should be used to create the thin logical volumes.

>     - We will have multiple thin pools per volume group. Multiple device classes can be mapped to a single volume group where each device class is mapped to a different thin pool within that volume group.

Reviewer 5: ```suggestion
    - We will support multiple thin pools per volume group by having multiple device classes mapping to a single volume group where each device class is mapped to a different thin pool within that volume group.
```

- Command to be used for montoring:
``` go=
   lvs
   LV          VG   Attr       LSize   Pool  Origin Data%  Meta%  Move Log Cpy%Sync Convert
   lv1         vg1  -wi-ao---- 600.00m
   thin_volume vg1  Vwi-a-tz--   1.00g tpool        3.20
   tpool       vg1  twi-aotz-- 200.00m              16.38  1.37
```
- Topolvm will only monitor the thin-pools provided in the lvmd.yaml config file.

#### Capacity awareness
- Topolvm uses capacity awareness to ensure that new pod is created on the node where available capacity of the volume group is larger than the requested capacity.`capacity.topolvm.cybozu.com/<device-class>` annotation is added on each node which consists of the available capacity on the volume group.
- Creation of thin logical volumes doesn't update the available device class capacity until the user actually uses storage on the thin volumes.
- So we need to ensure that all the thin logical volumes don't get created on the same node.
- This can be done by using the `overprovisionRatio`.
- Value of `capacity.topolvm.cybozu.com/<device-class>_mlog<thin-pool>` annotation will be calculated as:
  `(thin-pool-size * overprovisionedRatio)-(sum-of-all-thin-lv-sizes-in-thin-pool)`

> - Available thin pools on the volume groups will be obtained from lvmd.yaml config file.

Reviewer: Let's consider below example for a single node to read in conjunction with above responses:

## Setup:
1. Sample lvmd config with thin-pool in an existing volume group `vg0`
``` yaml
device-classes:
  - name: ssd
    volume-group: vg0
    default: true
    overprovision-ratio: 10.0
    thin-pools:
        -name: pool0
```
2. Total size of `vg0` is 50G and 10G is already used for thick volumes
3. We (~admin) created `vg0/pool0` of size 10G and with overprovision-ratio of `10.0` we respect thin-lv creations amounting upto 10.0 X 10G = 100G virtual size
- Now the annotations on this node will be as follows:
```
capacity.topolvm.cybozu.com/ssd: 30G
capacity.topolvm.cybozu.com/ssd_mlogpool0: 100G
```
30G --> 50G (vg_size) - 10G (thick-lvs) - 10G (thin-pool)
100G --> 100G virtual size as mentioned above since there's no thin-lv created on this pool yet

## Request:

a. thick lv creation:

1. below params exists on storage class (sc1) for creation of thick volumes from a specific device class
```
topolvm.cybozu.com/device-class: ssd
```
2. pvc request has come up on this sc1 for 5G and topolvm created it and updated metrics & annotations
```
capacity.topolvm.cybozu.com/ssd: 25G
capacity.topolvm.cybozu.com/ssd_mlogpool0: 100G (same no change)
```

b. thin lv creation:
1. below params exists on storage class (sc2) for creation of thin volumes from a specific device class & pool
```
topolvm.cybozu.com/device-class: ssd
topolvm.cybozu.com/thin-pool-name: pool0
```
2. pvc request has come up on this sc2 for 50G and topolvm created it and updated metrics & annotations
```
capacity.topolvm.cybozu.com/ssd: 25G (same no change)
capacity.topolvm.cybozu.com/ssd_mlogpool0: 50G
```
50G --> 100G (size with overprovision) - 50 G (which is just created)
3. another pvc request on sc2 for 30G and below will be the updated annotations
```
capacity.topolvm.cybozu.com/ssd: 25G (same no change)
capacity.topolvm.cybozu.com/ssd_mlogpool0: 20G
```
20G --> 100G (size with overprovision) - 50 G (existing) - 30G (current)

> - Available thin pools on the volume groups will be obtained from lvmd.yaml config file.

Reviewer 3: @leelavg 
Thank you for sharing your idea. I understand. I reply after discussion Open Questions.

## Open Questions
- Should topolvm create the thin-pools instead of the admin/operator?

> - If multiple thin-pools are supported for each volume group, then how to decide which thin-pool should be used to create the thin logical volumes.

Author: > It's difficult to answer the question. I don't expect multiple thin-pools for each VG. I'd like to support only one node per one VG.
I think you meant one thin-pool per VG ?

> - If multiple thin-pools are supported for each volume group, then how to decide which thin-pool should be used to create the thin logical volumes.

Reviewer: > BTW, there is some code about thin-pool, but, currently, there code is not used yet. So please revise them as you need when you implement this design.
- ack, we (I'm with @sp98) did a poc for a similar design and re-used the existing code and it was working for the design

> - If multiple thin-pools are supported for each volume group, then how to decide which thin-pool should be used to create the thin logical volumes.

Reviewer 3: @sp98 
> I think you meant one thin-pool per VG ?

Sorry, yes. I meant one thin-pool per VG.

> Do you think in future we might have to support more than 1 thin-pools per volume Group?

No, I don't. I don't want to support more than 1 thin-pools per volume Group.

> - If multiple thin-pools are supported for each volume group, then how to decide which thin-pool should be used to create the thin logical volumes.

Author: Based on the latest discussions, we will be supporting multiple thin pools per volume group. Marking this conversation as resolved.

    - User should decide which thin-pool to use. This can done via a StorageClass parameter.

> - How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.

Author: > The capacity aware scheduling code will be complicated very much if there are normal pools and thin pools at the same time.

Agree that presence of both thin and thick pools will complicate the scheduling decisions. 

> In addition, it's difficult to create practical logic in such case. So, how about adding a description to prohibit the co-existence of normal pool and thin pool in docs/limitations.md?

Do you mean that entire topolvm cluster can only be created either using thin-pools or thick pool and not with a combination of both?   If we do this, then existing topolvm customers won't be able to make use of the thin volumes feature. This will be a big disadvantage.

> - How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.

Reviewer: @cupnes
- the idea here is to have distinct annotations for thin-pool and remaining space of VG (current method) which can be used to take scheduling decisions based on the ask for thin/thick logical volume

> - How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.

Reviewer 3: @sp98 
> Do you mean that entire topolvm cluster can only be created either using either thin-pools or thick pools and not with a combination of both?

No, I don't. I mean that one volume group is for either thin volume or thick volume. VGs for thin volumes and VGs for thick pools can be in same TopoLVM cluster.

@leelavg 
I understood.

> - How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.

Author: > No, I don't. I mean that one volume group is for either thin volume or thick volume. VGs for thin volumes and VGs for thick pools can be in same TopoLVM cluster.

Thanks for the clarification @cupnes . I see you have acknowledged one of the [comments](https://github.com/topolvm/topolvm/pull/442#discussion_r822538501) provide by my colleague @leelavg about how we can use both thin and thick volumes on the same Volume group. 

Do you see any complexities for having both thick and thin pools on the same VG? We would like to have this because it will help existing customers using thick volumes to create thin-pools and create thin-volumes on them.

> - How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.

Reviewer 3: @sp98 Thank you for additional information. I understood the need. Sorry, I will reply on this discussion next week.

> - How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.

Reviewer 6: @cupnes - in particular we would like to keep existing thick-LVs on the VG, then add a thin-pool and from that point onwards only add thin-LVs. Thus the scheduling algorithm wouldn't need to consider this VG any more for new thick-LVs.

> - How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.

Reviewer 7: Is there really a need for thin and thick to coexist in the same VG, or is it more about trying to help clusters upgrade and starting using thin after they had been created with thick? The thin/thick modes are so different, my first impression is that VGs should really be dedicated to either thick or thin. Instead of trying to support both in the same VG, perhaps there is a simpler migration path to convert thick VGs to thin VGs, or truncate unused space in the thick VG to create a new thin VG.

> - How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.

Author: > ... or is it more about trying to help clusters upgrade and starting using thin after they had been created with thick? 

That's the only use case for us AFAIK.

> - How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.

Author: The resolution is to have multiple device classes per volume group and each device class is mapped to a single thin pool on that volume group.  Marking this conversation as resolved.

> - Should we expect auto-extend of thin-pool to be enabled by default? What happens when the user has already enabled the autoextend feature when doing the monitoring?

Author: My bad. The second part of this question should have been added in the next question.  So the problem is, 

``` if auto extend is enabled by the user, the thin-pool size will extend automatically. How will the topolvm track this update in size and update the related metircs accordingly. Should topolvm constantly poll the available capacity of the thin pools? Should that be a performance concern? ```

> - Should we expect auto-extend of thin-pool to be enabled by default? What happens when the user has already enabled the autoextend feature when doing the monitoring?

Reviewer 3: @sp98 
> My bad. The second part of this question should have been added in the next question. So the problem is,
> `if auto extend is enabled by the user, the thin-pool size will extend automatically. How will the topolvm track this update in size and update the related metrics accordingly. Should topolvm constantly poll the available capacity of the thin pools?

I understood. My answer for the problem is follow.
https://github.com/topolvm/topolvm/pull/442#discussion_r821356000

> Should that be a performance concern?`

No, it is not a performance concern. The polling period of `checkFunc()` is 1 minute.
https://github.com/topolvm/topolvm/blob/e00d2f66180b9058e3c78268ce095de34805291f/pkg/topolvm-node/cmd/run.go#L85

> - Does topolvm constantly poll the available capacity of the thin pools?

Reviewer: - We would like have an opinion from you about pooling for thin pool size for every minute?
- Is it fine if we do not implement this now and think of other ways if possible?

> - Does topolvm constantly poll the available capacity of the thin pools?

Reviewer 2: > We would like have an opinion from you about polling for thin pool size for every minute?

It's desirable because it's the simplest solution.

> Is it fine if we do not implement this now and think of other ways if possible?

It depends on the reason. Are there any reasons that you don't want to do so? IMO, it's not so difficult to implement and it can avoid the problem that anyone outside K8s uses thin pools written in device classes.

> - Does topolvm constantly poll the available capacity of the thin pools?

Reviewer: @satoru-takeuchi / @cupnes I was awaiting initial round of reviews on implementation to discuss on pending items however expanding this conversation itself to think about it in parallel.

Problem Statement:
- When the thin pool is expanded by admin topolvm will not realise the increase in size instantly unless an operation is performed which sends `s.notify()`

Solution that's discussed here:
- Use the existing mechanism to get thin pool size and update the node annotation every minute using `checkFunc` and `NewChecker`

Concerns:
- currently annotation and monitoring updates are triggered by `notify` event for all the device-classes
- we can't target a single device-class annotation be updated
> It's desirable because it's the simplest solution.
- by above, you mean the solution is straightforward, if yes, could you pls expand your answer a little bit about how the components should interact to achieve this?

My thoughts:
- we could just document to wait for 10 min after expanding the thin pool to realise the new size as I could see a ticker sending `notify` every 10 min here https://github.com/topolvm/topolvm/blob/f76b553b2d0bdf2ac71d009a4cd0ec36cd9aea5d/pkg/lvmd/cmd/root.go#L112-L123
- pls correct me if above is for something else entirely

> - Does topolvm constantly poll the available capacity of the thin pools?

Reviewer 3: We will reply next week.

> - Does topolvm constantly poll the available capacity of the thin pools?

Reviewer 2: @leelavg I'll get back to you later on @cupnes 's behalf because he's extremely busy right now with another jobs.

> - Does topolvm constantly poll the available capacity of the thin pools?

Reviewer: - ack, noted. Meanwhile, a review on implementation PR would also be helpful 😅

> - Does topolvm constantly poll the available capacity of the thin pools?

Reviewer: > I don't think it will be a serious problem. Please let me know if you have any concern.
- now it's all good, what I meant earlier was, this particular function was invoking `getlvbytes` on a default device class to verify lvm subsytem is functional  as a means of readiness checks
- using this functionality to update the node annotation is a little bit contrived

discussion on this thread is finalised, this can be resolved IMHO.

> - Does topolvm constantly poll the available capacity of the thin pools?

Reviewer 4: @leelavg
[satoru-takeuchi](https://github.com/satoru-takeuchi) is urgent today and I respond this question.

> now it's all good, what I meant earlier was, this particular function was invoking getlvbytes on a default device class to verify lvm subsytem is functional as a means of readiness checks
>
> using this functionality to update the node annotation is a little bit contrived

As you recognize. We don't need to modify around `checkFunc` function. We had to only modify `vgService.send` to update node labels of pool capacity. Could you please update the answer.

- TopoLVM updates the node's annotation every 10 mins using `notify` function as a trigger as you know.
- The annotations are updated on `metricsExporter.updateNode` by values get using `VGServece.Watch`.
- Modify `vgService.send` to update value used by `VGServece.Watch`.

> discussion on this thread is finalised

I agree with you.

> - Does topolvm constantly poll the available capacity of the thin pools?

Reviewer: > Modify vgService.send to update value used by VGServece.Watch.
- this is already part of the implementation, I'll add how metrics & annotations are being updated to this design doc.
